### ISSUE TITLE
HRV Measures

### ISSUE THREAD

**Issue by amtriathlon** [GoldenCheetah/GoldenCheetah#2568]
I would like to discuss how to add HRV Measures for 3.5, what follows is what I plan to do if there are no objections, better ideas or others working on this:
1) Introduce HRV Measures is a similar way to Body Measures, day/time record with: HR,AVNN,SDNN,rMSSD,pNN50,LF,HF and LFHF. Cached on Athlete, API to get/set.
2) Framework for Download copied from BodyMeasures, no online services yet.
3) Import from CSV file in HRV4Training format with unsupported columns and empty rows removed, supported columns are: timestamp_measurement, HR, AVNN, SDNN, rMSSD, pNN50, LF, HF, HRV4T_Recovery_Points. timestamp_measurement and rMSSD are mandatory.
4) Add pseudo metrics to access HRV measures
5) HRV Metrics integration: add an option to FilterHRV dataprocessor to set HRV Measures from computed HRV Metrics
6) UI to edit HRV Measures in Options/Preferences
7) Start a wiki page with how to use information:
 (https://github.com/GoldenCheetah/GoldenCheetah/wiki/HRV)
8) Upload a chart with daily and rolling averages values for the most commonly used measures. Uploaded to the cloud.

All items above are implemented in master and available on  [v3.5-DEV1708](https://github.com/GoldenCheetah/GoldenCheetah/releases/tag/v3.5-DEV1708).
For other features which also apply to BodyMeasures s.t. plot by days in LTM charts (ala PMC) and access from formulas, R and REST interfaces probably would be better to open other FRs.

**Comment by liversedge** [comment]
Do we need to treat measures like we do pmc/stress calculations, when you request them a series is created for every day (not ride) ?

Additionally, RideItem::getMeasure(..,) like RideItem::getWeight() ?

The rest sounds perfect.

**Comment by amtriathlon** [comment]
>Additionally, RideItem::getMeasure(..,) like RideItem::getWeight() ?
Yes, I will include this in part 1.

> Do we need to treat measures like we do pmc/stress calculations, when you request them a series is created for every day (not ride) ?
I was thinking more in the line of Athlete Weight, the pseudo metrics of part 4 use getMeasure to obtain the HRV measurement closer to the date of the ride, mostly for implementation simplicity.

**Comment by liversedge** [comment]
Sure, I was thinking about plots with daily points, not just on the day of a ride like with PMC.

**Comment by amtriathlon** [comment]
Agree, to be able to plot measures independent of rides would be a nice feature to have, also to add measures of sleep, perceived well being, recovery, etc.

**Comment by ninardos** [comment]
I would have some ideas and comments on what I would deem useful regarding HRV tracking in GC. I´m not sure this is the right place to put it, but couldn´t find another one.
1. Analyse a single HRV reading: 
a) plot: raw (all measured values) and filtered (only the ones supposed to be correct, i. e. excluding artifacts)
b) metrics (the standard statistically base metrics)
c) rating of recovery status, maybe on a scale from 1-10 or from deep red to bright green - sympathetic / parasympathetic), that takes into account recent changes (maybe 7 days) and long term basis (2 months)

2. Plot metrics and trends for HRV morning readings over time. 
a) maybe this could be realized by providing a Chart that only plots certain tagged activities.
b) possibility of integration of the HRV metrics in the PMC type Charts.
c) apart from the metrics themselves, rolling averages of e. g. reting HR, HRV, etc. for short term (7 day) and long term 1-2 months) should be available

3. Import
As most people measure HRV with an app today, a sort of batch import  / export or more conveniently a interface to the most frequently used apps would be needed.
a) easy import of a single reading (e. g. .csv file with date, time and measured R-R values)
b) interface to EliteHRV, ithlete, HRV4Training

Hope this is the right place for this kind of questions - if not, please let me know.

Best regards,

Stefan

**Comment by rgeorge1985** [comment]
Stefan,

Nice thoughts, just a few interjections...

1 a) In my opinion, it would be best if R-R time series correction is left
to the user...and to a specialized, established software of choice, NOT GC.
Such software will have proper documentation on what methodology it uses to
correct the R-R signal, because there can be many errors to look for and
advanced mathematical techniques to deal with those errors. Ideally,
developing these should not be the burden of GC developers, why re-invent
the wheel, and in the process, potentially making a flawed algorith that
yields incorrect and possibly misleading results to a user?

1 c) Such simplistic "colors" etc is based on the assumption that you can
reduce the interpretation IN ALL CASES to such simple conditions. In
reality, everyone's body is different. Such derived plots should be left to
the user to make on his own, not pre-programmed. Atleast from all my
research, there is no widely accepted scale or color scale or anything of
the sort backing this type of reductionism. All I see researchers having
developed are simple heuristics as helpful guidelines but they are not hard
coded rules.


2 c) I'm in support of 7 day rolling temporal metrics for rMSSD.


3. a) Importing should not be limited to just what apps throw out, it must
respect widely accepted file formats.  ASCII and EDF are two most popular,
but ASCII seems to be the least common denominator so GC can start with
that. I would hope that more than just to popular apps, GC fundamentally
has a basic import facility for files. For example, I don't pay a penny to
any app to do anything for me. I would hope GC does not favor one type of
user at the expense of another.

Just my two cents.



-Ron

**Comment by ninardos** [comment]
Hi Ron,

1c) I agree - there seems not to be the one and all solution. Yet it might be possible to use the known heuristics as far as reasonable to derive actionable advice. If not possible - of course not.

3 a) Import of some sort of ASCII / csv in documented format would be the first and most important solution. I agree that GC cannot possibly support all apps, so getting data into GC would have to be cared for by the interested user

1a) I personally think from a user perspective it is very valuable to have the SW take care of data quality. Kind of like the "find/correct outliers in power data, or recalculate speed from gps data". 
There are some methods that are consensus for correction. The most basic seems to work well, more advanced do not seem to improve overall quality (had a study on this, dont find it for reference now).
Correction of hrv RR data can be done based on two things: a) absolute limits (exclude everything above and below a threshhold - there could be reasonable defaults optionally set by the user, as everyone is different) b) exclude every measurement that is outside a window of +-20% of the last value / average of last n (n=3, 4, 5, TBD) values.

**Comment by ninardos** [comment]
A more general question that I have would be, how I could get more involved with contributing. I don´t have a C++ programming background, but some (maybe sufficient) background on e. g. HRV tracking and interpretation to get involved. Certainly I would be willing to help.

**Comment by amtriathlon** [comment]
Hi Stefan, most of these features are already implemented and will be available in the next development build for testing, you can test now if you build from source:

> Analyse a single HRV reading:
> a) plot: raw (all measured values) and filtered (only the ones supposed to be correct, i. e. excluding artifacts)

HRV readings are imported as XData from FIT files, there is a tool to filter (https://github.com/GoldenCheetah/GoldenCheetah/blob/a42adad4a0b5185d35ebf28bae31edaa8e39b526/src/FileIO/FilterHRV.cpp) and data can be charted as User Data in Activity Chart.

>b) metrics (the standard statistically base metrics)

The most common are implemented here: https://github.com/GoldenCheetah/GoldenCheetah/blob/75238f38685c18b4798489fe6e614a091686f4f2/src/Metrics/HrvMetrics.cpp

>c) rating of recovery status, maybe on a scale from 1-10 or from deep red to bright green - sympathetic / parasympathetic), that takes into account recent changes (maybe 7 days) and long term basis (2 months)

We are using a log transform of rMSSD for this purpose, similar to HRV4Training Recovery Points

>Plot metrics and trends for HRV morning readings over time.
>a) maybe this could be realized by providing a Chart that only plots certain tagged activities. 

HRV metrics are standard GC metrics and can be charted and filtered in Metric Trends charts, sample charts could be uploaded to the cloud service before release.
Additionally we plan to add the ability to use the Fix HRV tool to update HRV measurements at Rest, this is point 5 above, which is not implemented yet.

>b) possibility of integration of the HRV metrics in the PMC type Charts.

See above.

>c) apart from the metrics themselves, rolling averages of e. g. reting HR, HRV, etc. for short term (7 day) and long term 1-2 months) should be available

See above, rolling average is available as trend line and formula for all GC metrics.

>Import
>As most people measure HRV with an app today, a sort of batch import / export or more conveniently a interface to the most frequently used apps would be needed.
>a) easy import of a single reading (e. g. .csv file with date, time and measured R-R values)

This can be done using copy&paste on Editor tab.

>b) interface to EliteHRV, ithlete, HRV4Training

CSV import is documented here: https://github.com/GoldenCheetah/GoldenCheetah/blob/master/doc/user/hrvmeasures-format.txt, it was tested using files generated by HRV4Training but other should also work with some transformation.

**Comment by liversedge** [comment]
It would be nice to see body measure and HRV import as a single dialog, i.e. unified.
The import and share menus tend to proliferate and become unwieldy.

**Comment by rgeorge1985** [comment]
Ale,

Please provide some documentation for this FilterHRV subroutine. Would like to see whats in the black box. Are you borrowing techniques from another software, paper etc..?

**Comment by rgeorge1985** [comment]
Stefan

I would be happy to engage with you and contribute to the GC wiki on github. My perception is that a little more could be done in the documentation department. Maybe a basic writeup to HRV and throw some good well cited research work into it? There must a method to the madness, HRV needs a user guideline      for sure! What does Mark, Ale think ?

**Comment by leifwar** [comment]
Most of the code is based on https://physionet.org/tutorials/hrv-toolkit.

**Comment by rgeorge1985** [comment]
Leif,

Noted, this is good, no re-invention of wheel, re-use what has been
validated.

Are you attributing the Physionet source in the code? Just curious how
you'll take care of that within GC.

-Ron

**Comment by rgeorge1985** [comment]
Ale,

I read the csv import documentation, I note the depedance on HRV4Training,
not something I use so concerned about the headers...
timestamp_measurement,HR,AVNN,SDNN,rMSSD,pNN50,LF,HF,HRV4T_Recovery_Points

1) Will a Polar R-R text file work (downloaded off Flow)? These text files
have no header information and I think the data contains one column with
intervals. Very amenable to import into Kubios etc.

2) If no to 1), what should one to do get the file into the format this
import functionality requires.

3) How could one start evaluation of these features, primarily import,
filterHRV etc etc . I have dev3.5 installed.

Ron

**Comment by liversedge** [comment]
@rgeorge1985 if you want to write up some science based documentation on using HRV we can add it to the science section of the goldencheetah website.

I would really like to add a section on training planning and tracking using metrics that have been validated in the literature. HRV would be a little part of that.

**Comment by amtriathlon** [comment]
@rgeorge1985 what I recently added is what is detailed in my first comment, the import deals with HRV metrics at rest already processed by another source, I tested it using HRV4Training but the source can be anything, including Kubios. This will be available in the next development build.

The ability to import R-R records is available for Garmin, Suunto and Polar devices in their native formats (FIT, SML and HRM/PDD resp.) in V3.5-DEV1705 together wiith Edit>Filter HRV tool and the HRV metrics Leif has incorporated.
